config/initializers/at_exit.rb

```diff
@@ -0,0 +1,61 @@
+#
+# Graceful shutdown of web workers
+#
+# Note this is executed in a different context, Ruby on Rails application is
+# not available, this includes Rails.env, Rails.logger, Rails.root and others.
+#
+# Signal.trap cannot be used as it is redefined in web server (Puma). Multiple
+# at_exit blocks are allowed.
+#
+
+SHUTDOWN_LOGFILE = File.expand_path(File.dirname(__FILE__) + "../../../log/shutdown.log")
+PROMETHEUS_DIR = File.expand_path(File.dirname(__FILE__) + "../../../tmp/prometheus")
+
+def logmsg(message)
+  File.open(SHUTDOWN_LOGFILE, 'a') { |f| f.puts "[#{Process.pid}] " + message }
+end
+
+def rotatelog
+  return unless File.exist?(SHUTDOWN_LOGFILE)
+  return if File.size(SHUTDOWN_LOGFILE) < 1_000_000
+  FileUtils.rm_f(SHUTDOWN_LOGFILE)
+end
+
+def check_pid(pid)
+  Process.getpgid(pid.to_i)
+  true
+rescue Errno::ESRCH
+  false
+end
+
+def cleanup_prometheus
+  # delete my own files
+  logmsg "Deleting #{PROMETHEUS_DIR}/*#{Process.pid}.bin"
+  FileUtils.rm_f(Dir.glob("#{PROMETHEUS_DIR}/*#{Process.pid}.bin"))
```

Comment:
If you're deleting `*$pid`, you will also delete unrelated files, right? Let's say you have pid 10 that exits, but pid 210 is still running. This will match both of them.

Reply:
Yes this is an attempt to "cleanup" this temporary directory. All the processes which are no longer active left some data which only makes things slow.

We actually do a full cleanup of this directory during startup (delete all files), so this could work fine without this extra "dead cleanup". I can remove this if needed.

Reply:
+1 for removing this